api/src/main/java/edu/wpi/first/shuffleboard/api/sources/recording/Serialization.java

```diff
@@ -83,35 +108,165 @@ public static void saveRecording(Recording recording, Path file) throws IOExcept
+      raf.seek(pos);
+      final int len = raf.read(sourceIdBuffer, 0, sourceIdLength);
+      String sourceId = new String(Arrays.copyOfRange(sourceIdBuffer, 0, len), StandardCharsets.UTF_8);
+      pos += len;
+      constantPoolSize += len;
+      sourceIds.add(sourceId);
+    }
+    byte[] newConstantPoolEntries = recording.getData().stream()
+        .map(TimestampedData::getSourceId)
+        .distinct()
+        .filter(id -> !sourceIds.contains(id))
+        .map(Serialization::toByteArray)
+        .reduce(new byte[0], Bytes::concat);
+    recording.getData().stream()
+        .map(TimestampedData::getSourceId)
+        .distinct()
+        .filter(s -> !sourceIds.contains(s))
+        .forEachOrdered(sourceIds::add);
+    byte[] newBodyEntries = recording.getData().stream()
+        .map(data -> {
+          final DataType type = data.getDataType();
+          final Object value = data.getData();
+
+          final byte[] timestamp = toByteArray(data.getTimestamp());
+          // use int16 instead of int32 -- 32,767 sources should be enough
+          final byte[] sourceIdIndex = toByteArray((short) sourceIds.indexOf(data.getSourceId()));
+          final byte[] dataType = toByteArray(type.getName());
+          final byte[] dataBytes = encode(value, type);
+
+          return Bytes.concat(timestamp, sourceIdIndex, dataType, dataBytes);
+        })
+        .reduce(new byte[0], Bytes::concat);
+
+    if (newConstantPoolEntries.length > 0) {
+      // Update the number of source IDs
+      int totalNumSourceIds = sourceIds.size();
+      raf.seek(Offsets.CONSTANT_POOL_HEADER_OFFSET);
+      raf.writeInt(totalNumSourceIds);
+
+      // Update the constant pool
+      int constantPoolEnd = Offsets.CONSTANT_POOL_HEADER_OFFSET + SIZE_OF_INT + constantPoolSize;
+      insertDataIntoFile(file, constantPoolEnd, newConstantPoolEntries);
+    }
+    Files.write(file, newBodyEntries, StandardOpenOption.WRITE, StandardOpenOption.APPEND);
+  }
+
+  /**
+   * Inserts binary data into a file at a specific position.
+   *
+   * @param path the path to the file to insert into
+   * @param pos  the position in the file to insert the data at
+   * @param data the data to insert
+   *
+   * @throws IOException if the file could not be modified
+   */
+  static void insertDataIntoFile(Path path, long pos, byte[] data) throws IOException {
+    if (Files.notExists(path)) {
+      throw new NoSuchFileException("The file specified does not exist: " + path);
+    }
+    if (!Files.isRegularFile(path, LinkOption.NOFOLLOW_LINKS)) {
+      // Make sure it's a normal file (not a directory and not a symlink/shortcut to another file)
+      throw new IllegalArgumentException("Cannot insert into " + path);
+    }
+    Objects.requireNonNull(data, "Data cannot be null");
+    if (data.length == 0) {
+      // Nothing to insert, do nothing
+      return;
+    }
+    Path fileName = path.getFileName();
+    if (fileName == null) {
```

Comment:
Have to do this check to appease FindBugs. This will never be null for a file, which we check for above

Reply:
What FindBugs warning is being thrown?

Reply:
It was complaining about a possible null dereference. I don't remember the exact warning